core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushMergeWriterUpdateIntoConnector.java

```diff
@@ -0,0 +1,151 @@
+import static io.trino.sql.planner.plan.Patterns.source;
+import static io.trino.sql.planner.plan.Patterns.tableFinish;
+import static io.trino.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+/**
+ *  This version support only constant updates
+ *  and fall back to default behaviour in all other cases
+ */
+public class PushMergeWriterUpdateIntoConnector
+        implements Rule<TableFinishNode>
+{
+    private static final Capture<MergeWriterNode> MERGE_WRITER_NODE_CAPTURE = newCapture();
+    private static final Capture<MergeProcessorNode> MERGE_PROCESSOR_NODE_CAPTURE = newCapture();
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+    private static final Capture<ProjectNode> PROJECT_NODE_CAPTURE = newCapture();
+
+    private static final Pattern<TableFinishNode> PATTERN =
+            tableFinish().with(source().matching(
+                    mergeWriter().capturedAs(MERGE_WRITER_NODE_CAPTURE).with(source().matching(
+                            mergeProcessor().capturedAs(MERGE_PROCESSOR_NODE_CAPTURE).with(source().matching(
+                                    project().capturedAs(PROJECT_NODE_CAPTURE).with(source().matching(
+                                            tableScan().capturedAs(TABLE_SCAN)))))))));
+
+    public PushMergeWriterUpdateIntoConnector(PlannerContext plannerContext, TypeAnalyzer typeAnalyzer, Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.typeAnalyzer = requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+    }
+
+    private final Metadata metadata;
+    private final PlannerContext plannerContext;
+    private final TypeAnalyzer typeAnalyzer;
+
+    @Override
+    public Pattern<TableFinishNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TableFinishNode node, Captures captures, Context context)
+    {
+        MergeWriterNode mergeWriter = captures.get(MERGE_WRITER_NODE_CAPTURE);
+        MergeProcessorNode mergeProcessor = captures.get(MERGE_PROCESSOR_NODE_CAPTURE);
+        ProjectNode project = captures.get(PROJECT_NODE_CAPTURE);
+        TableScanNode tableScan = captures.get(TABLE_SCAN);
+
+        List<String> orderedColumnNames = mergeWriter.getTarget().getMergeParadigmAndTypes().getColumnNames();
+        List<? extends Node> mergeAssignments = project.getAssignments().get(mergeProcessor.getMergeRowSymbol()).getChildren();
+
+        ImmutableMap.Builder<String, Constant> assignmentsBuilder = ImmutableMap.builder();
+
+        for (int i = 0; i < orderedColumnNames.size(); i++) {
+            String columnName = orderedColumnNames.get(i);
+            Node assigmentNode = mergeAssignments.get(i);
+            if (!isSymbolName(assigmentNode) && !isConstantUpdateValue(assigmentNode)) {
```

Comment:
Can you remove `!isSymbolName(assigmentNode)` condition from here? You support constant only anyway.

Reply:
I think, because merge row looks like this:
For example we have table: `int col1, int col2, int col3.`
If we execute query `UPDATE tableName SET col2 = 2;`
Merger row will be looking like this : `Symbol(col1), Constant(2), Symbol(col3)`

So for columns, which are not updated I will get symbol(column name).
And we need just skip such elements.

Reply:
Ok, so I reformat this code (loop body) as below

```
if (isSymbol(assigmentNode)) {
   // the columns is not updated
   continue;
}
if (! isConstantUpdateValue) {
   return Result.empty();
}   

ConnectorExpressionTranslator.ConnectorExpressionTranslation expressionTranslation = ConnectorExpressionTranslator.translateConjuncts(
vlad-lyutenko marked this conversation as resolved.
                        context.getSession(),
                        ((Expression) assigmentNode),
                        context.getSymbolAllocator().getTypes(),
                        plannerContext,
                        typeAnalyzer);

ConnectorExpression connectorExpression = expressionTranslation.connectorExpression();

if (!(connectorExpression instanceof Constant)) {
      return Result.empty();
}
assignmentsBuilder.put(columnName, (Constant) connectorExpression);
```

Reply:
refactored